src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,690 @@
+type SpansById = Map<number, SpanWithChildren>;
+
+/**
+ * Extrinsic indexs decoded from storage read events, keyed by the span Id to which
+ * the original storage event belonged to.
+ */
+type ExtrinsicIndexBySpanId = Map<number, BN>;
+
+/**
+ * Mapping of span id => array of the events belonging to the span.
+ */
+type EventsByParentId = Map<number, ParsedEvent[]>;
+
+/**
+ * Mapping of address => `system::Account` events that belong to the address
+ */
+type AccountEventsByAddress = Map<string, ParsedAccountEvent[]>;
+
+/**
+ * Class for processing traces from the `state_traceBlock` RPC endpoint.
+ *
+ * Assumptions:
+ * - Spans do not start in sorted order.
+ * - Events start in sorted order.
+ *
+ * For a conceptual overview on traces from block excution and `Operation` construction
+ * consult [this diagram.](https://docs.google.com/drawings/d/1vZoJo9jaXlz0LmrdTOgHck9_1LsfuQPRmTr-5g1tOis/edit?usp=sharing)
+ */
+export class Trace {
+	/**
+	 * Known storage keys.
+	 */
+	private keyNames;
+	constructor(
+		api: ApiPromise,
+		/**
+		 * `state_traceBlock` RPC response `result`.
+		 */
+		private traceBlock: BlockTrace,
+		/**
+		 * Type registry corresponding to the exact runtime the traces are from.
+		 */
+		private registry: Registry
+	) {
+		if (!traceBlock.spans.length) {
+			throw new InternalServerError(
+				'No spans found. This runtime is likely not supported with tracing.'
+			);
+		}
+
+		this.keyNames = Trace.getKeyNames(api);
+	}
+
+	// In the future this will need a more chain specific solution that creates
+	// keys directly from metadata. https://github.com/paritytech/substrate-api-sidecar/issues/528
+	/**
+	 * Get all the storage key names based on the ones built into `api`.
+	 *
+	 * @param api ApiPromise
+	 */
+	private static getKeyNames(api: ApiPromise): Record<string, KeyInfo> {
+		const moduleKeys = Object.keys(api.query).reduce((acc, mod) => {
+			Object.keys(api.query[mod]).forEach((item) => {
+				const queryObj = api.query[mod][item];
+				let key;
+				try {
+					// `slice(2)` is to slice off '0x' prefix
+					key = queryObj.key().slice(2);
+				} catch {
+					key = queryObj.keyPrefix().slice(2);
+				}
+
+				acc[key.toString()] = {
+					module: mod,
+					item,
+				};
+			});
+
+			return acc;
+		}, {} as Record<string, KeyInfo>);
+
+		return { ...moduleKeys, ...WELL_KNOWN_KEYS };
+	}
+
+	/**
+	 * Extract extrinsic indexes from all `:extrinsic_index` storage item read events.
+	 *
+	 * Note: Expects the given `spanIds` are from an apply extrinsic action group because
+	 * apply extrinsic action groups should contain read events to `:extrinsic_index`.
+	 *
+	 * @param spanIds List of spanIds for an apply extrinsic action group
+	 * @param extrinsicIndexBySpanId Map of spanId => extrinsicIndex
+	 */
+	private static extractExtrinsicIndex(
+		spanIds: number[],
+		extrinsicIndexBySpanId: ExtrinsicIndexBySpanId
+	): BN {
+		// There are typically multiple reads of extrinsic index when applying an extrinsic,
```

Comment:
so I don't really understand why you have to `reduce` to a list when it's an error having multiple intrinsic numbers,

Thus, can't just assign that single variable that you check when iterating over the `spanIds`?

For example:

```
let uniq = undefined;
for (let id of spanIds) {

    if (uniq === undefined) {
       uniq = id;
    } 
    
    if (uniq != id) {
       throw new InternalServerError(
       'Expect extrinsic to only be applied at a single index.'
        );
    }
}
```

Reply:
Nice, yeah that is a good point. Re-wrote it so its still a reduce, but now we just check the value is either undefined or matches the previous value